libvast/vast/bloom_filter.hpp

```diff
@@ -65,11 +70,25 @@ class bloom_filter : detail::equality_comparable<
 
   /// Adds an element to the Bloom filter.
   /// @param x The element to add.
+  /// @returns `void` if the tracking policy is `tracking::no`,
+  /// otherwise a `bool` that is `false` iff *x* already exists in the filter.
   template <class T>
-  void add(T&& x) {
+  auto add(T&& x) {
     auto& digests = hasher_(std::forward<T>(x));
-    for (size_t i = 0; i < digests.size(); ++i)
-      bits_[position(i, digests[i])] = true;
+    if constexpr (tracking_policy == policy::tracking::no) {
+      for (size_t i = 0; i < digests.size(); ++i)
+        bits_[position(i, digests[i])] = true;
+    }
+    if constexpr (tracking_policy == policy::tracking::yes) {
+      auto unique = false;
+      for (size_t i = 0; i < digests.size(); ++i) {
+        auto bit = bits_[position(i, digests[i])];
+        if (!bit)
+          unique = true;
+        bit = true;
+      }
```

Comment:
Do we really need a separate template policy argument **and** duplicated code for such simple functionality? Why not just always return a (discardable) bool here?

Reply:
I thought about this as well, but then felt uneasy with the extra branching. Granted, I'm not sure if the bandwidth on the Bloom filter construction site matters.

What would you prefer?

Reply:
I would be extremely suprised if that turns out to be measurable. The old value needs to be loaded + stored anyways to apply the masking, so the only additional work is an  `or` operation (you could write it as `unique |= (byte == 0)` if you don't trust the compiler), which should be amongst the fastest possible things you can do to a register.

It's not really important, so if you have other things to do I don't mind merging this as it is, just be prepared that there's a high chance that someone will remove the extra code again in the future for better maintainability ;)

Reply:
I'm nodding along, this all makes sense. 

I've removed the tracking policy, but kept the "new-style" enum class policy refactoring.